### PR TITLE
Basic record-replay

### PR DESCRIPTION
**Pull request by digetx**
For the starter there is only DRM record-replaying and that's it, but I already found it quite useful. Replaying of grate tests and opentegra works.

It would be nice to have ability to easily modify recorded data and display individual render targets / textures, that's something to be done later.

### CODE REVIEW

**Comment by digetx** [comment]
### Recording

Recording is enabled using `LIBWRAP_RECORD_PATH=/path/recfile` environment variable, in result recfile contains the recorded data that can be replayed.

### Replaying

Example:

	grate/tools/replay --recfile /path/recfile 

Use Space key to pause execution and to single-step to next job submit.

File: tools/replay.c [hunk 1]
```diff
@@ -0,0 +1,830 @@
...
+			break;
+
+		default:
+			fprintf(stderr, "Invalid arguments\n\n");
+			abort();
+		}
+	} while (c != -1);
+
+	if (!path) {
+		fprintf(stderr, "'--record path' is missing\n\n");
```
Comment (kusma) [inline tools/replay.c]: I think this should be `--recfile` instead of `--record`, no?

File: tools/replay.c [hunk 2]
```diff
@@ -0,0 +1,830 @@
...
+	assert(rbo != NULL);
+
+	return fread(rbo->map + page * 4096, 4096, 1, recfile);
+}
+
+static void set_bo_flags(unsigned int id, unsigned int ctx_id, uint32_t flags)
+{
+	struct rep_bo *rbo = lookup_bo(id, ctx_id);
+	assert(rbo != NULL);
+}
```
Comment (kusma) [inline tools/replay.c]: Isn't this supposed to do anything?

File: tools/Makefile.am [hunk 3]
```diff
@@ -3,7 +3,8 @@ noinst_PROGRAMS = \
 	cgc \
 	hex2float \
 	fp20 \
-	fx10
+	fx10 \
+	replay
```
Comment (kusma) [inline tools/Makefile.am:7]: I think this is in the wrong commit

File: tools/Makefile.am [hunk 4]
```diff
@@ -19,3 +20,11 @@ cgc_CPPFLAGS = \
...
 cgc_LDADD = \
 	../src/libcgc/libcgc.la
+
+replay_CPPFLAGS = \
+	-I$(top_srcdir)/include \
+	-I$(top_srcdir)/src/libhost1x \
+	-I$(top_srcdir)/src/libgrate
+
+replay_LDADD = \
+	../src/libgrate/libgrate.la
```
Comment (kusma) [inline tools/Makefile.am:30]: I think this is in the wrong commit

File: tools/meson.build [hunk 5]
```diff
@@ -3,12 +3,14 @@ tools = [
...
-	'fx10'
+	'fx10',
+	'replay'
 ]
 
 includes = include_directories(
 	'../../include',
-	'../../src/libgrate'
+	'../../src/libgrate',
+	'../../src/libhost1x'
```
Comment (kusma) [inline tools/meson.build:13]: I think this is in the wrong commit

File: src/libwrap/recorder.c [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#define _LARGEFILE64_SOURCE
+
+#include "recorder.h"
+#include "adler32.h"
```
Comment (kusma) [inline src/libwrap/recorder.c]: Can't we get adler32 from zlib distro-packages instead of shipping a copy of it?

File: src/libwrap/Makefile.am [hunk 7]
```diff
@@ -9,18 +9,22 @@ libgrate_wrap_la_CPPFLAGS += -DENABLE_RNN
...
 
 libgrate_wrap_la_SOURCES = \
+	adler32.h \
 	cdma_parser.c \
 	cdma_parser.h \
 	cgdrv.c \
 	disasm.h \
 	disasm.c \
 	drm.h \
+	drm_fourcc.h \
```
Comment (kusma) [inline src/libwrap/Makefile.am:20]: Doesn't look like this is the commit that adds drm_fourcc.h, is this in the wrong commit perhaps?

File: src/libwrap/meson.build [hunk 8]
```diff
@@ -1,16 +1,20 @@
 libwrap_sources =  files(
+	'adler32.h',
 	'cdma_parser.c',
 	'cdma_parser.h',
 	'cgdrv.c',
 	'disasm.c',
 	'disasm.h',
 	'drm.h',
+	'drm_fourcc.h',
```
Comment (kusma) [inline src/libwrap/meson.build]: Doesn't look like this is the commit that adds drm_fourcc.h, is this in the wrong commit perhaps?

**Review by kusma — COMMENTED** [review]
Very nice! There's a few minor issues here and there, but I really like this :)

File: tools/replay.c (hunk 1 above)
Comment (digetx) [reply tools/replay.c]: Yes!

File: tools/replay.c (hunk 2 above)
Comment (digetx) [reply tools/replay.c]: I'll add `TODO` comment here.

File: tools/Makefile.am (hunk 3 above)
Comment (digetx) [reply tools/Makefile.am:7]: Yes, that's what usually happens if I don't test each commit before pushing :)

File: src/libwrap/recorder.c (hunk 6 above)
Comment (digetx) [reply src/libwrap/recorder.c]: Seems.. yes. I think libwrap already indirectly depends on zlib, so good point!

File: tools/Makefile.am (hunk 4 above)
Comment (digetx) [reply tools/Makefile.am:30]: Yep

File: tools/meson.build (hunk 5 above)
Comment (digetx) [reply tools/meson.build:13]: Yep

File: src/libwrap/Makefile.am (hunk 7 above)
Comment (digetx) [reply src/libwrap/Makefile.am:20]: Yep

File: src/libwrap/meson.build (hunk 8 above)
Comment (digetx) [reply src/libwrap/meson.build]: Yep

**Comment by digetx** [comment]
Should be good now.

**Review by kusma — APPROVED** [review]
Very nice!

**Pull request merged**